nixos/modules/services/mail/dovecot.nix

```diff
@@ -1,19 +1,90 @@
-    mapAttrs' listToAttrs filter;
-  inherit (lib.strings) match;
+  inherit (lib) attrValues concatMapStringsSep concatStrings concatStringsSep
+    flatten imap1 isBool isDerivation isInt isList isPath isString
+    literalExpression mapAttrsToList mkEnableOption mkIf mkOption
+    mkRemovedOptionModule optional optionalAttrs optionalString singleton
+    types mkRenamedOptionModule nameValuePair mapAttrs' listToAttrs filter;
+  inherit (lib.generators) toPretty;
+  inherit (lib.lists) findFirstIndex length sublist reverseList;
+  inherit (lib.strings) escape match splitString;
 
   cfg = config.services.dovecot2;
   dovecotPkg = pkgs.dovecot;
 
   baseDir = "/run/dovecot2";
   stateDir = "/var/lib/dovecot";
 
+  /* Helper: Format string in way that may strech over multiple
+   * lines in a Dovecot configuration file
+   *
+   * This is a little complicated since Dovecot does not like when the last
+   * line is empty, so we first have to truncate all trailing empty lines
+   * before when can join them using the continuation line syntax.
+   */
+  formatDovecotMultiline =
+    string:
+      let
+        # Split input into line list
+        lines = splitString "\n" string;
+        # Find index of last non-empty line in line list
+        lastContentIndex = length lines - (findFirstIndex (
+          line: match "[[:space:]]*" line == null  # <=> Line is not empty
+        ) (length lines) (reverseList lines));
+        # Drop empty trailing lines from line list
+        trimmedLines = sublist 0 lastContentIndex lines;
+      in
+        # Reassemble content without trailing lines but with continuation markers
+        concatStringsSep " \\\n" trimmedLines;
+
+  /* Helper: Apply string escaping and quotation to Dovecot string if needed
+   *
+   * Dovecot is sometimes touchy about strings being quoted, so this only
+   * quotes the string when it is actually needed.
+   */
+  escapeDovecotString = string:
+    let
+      escapedString = escape ["\\" "\""] string;
+    in   if match "[[:space:]]*<.*|.*[[:space:]#\"\\].*|" string != null
+    then "\"${formatDovecotMultiline escapedString}\""
+    else string;
+
+  /* Helper: Convert a Dovecot value to a string
+   *
+   * Strings are escaped except for config variable interpolation, as such
+   * this function is unsuitable for performing in-Dovecot variable
+   * substitutions, however these can also easily be done using Nix instead.
+   *
+   * Reference: https://doc.dovecot.org/settings/types/
+   */
+  mkDovecotValue =
+    value: (
+      if isString value || isPath value || isDerivation value
+      then
+        escapeDovecotString (toString value)
+      else if isInt value
+      then
+        toString value
+      else if isBool value
+      then
+        if value then "yes" else "no"
+      else if isList value
+      then
+        concatMapStringsSep " " mkDovecotValue value
+      else
+        abort "mkDovecotValue: value not supported: ${toPretty {} value}"
```

Comment:
```suggestion
        throw "mkDovecotValue: value not supported: ${toPretty {} value}"
```

Reply:
Actually, abort works too, I suppose. I just see throw usually, so use whichever you want.

Reply:
Think I found abort in the docs somewhere, seen throw elsewhere too, I’ll change it.

Reply:
(Note for future readers: Difference between abort and throw is that throw can be caught during tryEval, while abort cannot, so you should generally use throw.)